File: proxy-protocol/src/main/java/com/mattunderscore/proxy/protocol/v1/PlainTextIPV4Deserialiser.java
/* Copyright © 2016 Matthew Champion
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:
 * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.
 * Redistributions in binary form must reproduce the above copyright
      notice, this list of conditions and the following disclaimer in the
      documentation and/or other materials provided with the distribution.
 * Neither the name of mattunderscore.com nor the
      names of its contributors may be used to endorse or promote products
      derived from this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL MATTHEW CHAMPION BE LIABLE FOR ANY
DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE. */

package com.mattunderscore.proxy.protocol.v1;

import com.mattunderscore.tcproxy.io.serialisation.AbstractByteBufferDeserialiser;
import com.mattunderscore.tcproxy.io.serialisation.AbstractByteBufferSerialiser;
import com.mattunderscore.tcproxy.io.serialisation.DeserialisationResult;
import com.mattunderscore.tcproxy.io.serialisation.Deserialiser;
import com.mattunderscore.tcproxy.io.serialisation.NotDeserialisableResult;

import java.net.Inet4Address;
import java.net.InetAddress;
import java.net.UnknownHostException;
import java.nio.ByteBuffer;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Deserialiser for {@link Inet4Address} from {@link ByteBuffer}.
 * @author Matt Champion on 20/05/16
 */
public final class PlainTextIPV4Deserialiser extends AbstractByteBufferDeserialiser<InetAddress> {
    public static final Deserialiser<InetAddress, ByteBuffer> INSTANCE = new PlainTextIPV4Deserialiser();
    private static final Pattern IPV4_PATTERN = Pattern.compile(
            "([0-9][0-9]?[0-9]?)\\.([0-9][0-9]?[0-9]?)\\.([0-9][0-9]?[0-9]?)\\.([0-9][0-9]?[0-9]?)");

    private PlainTextIPV4Deserialiser() {
    }

    @Override
    protected Result<InetAddress> doRead(ByteBuffer buffer) {
        final byte[] bytes = new byte[buffer.remaining()];
        buffer.get(bytes);
        final String address = new String(bytes);

        final Matcher matcher = IPV4_PATTERN.matcher(address);
        if (matcher.matches()) {
            final int firstByte = Integer.parseInt(matcher.group(1), 10);
            final int secondByte = Integer.parseInt(matcher.group(2), 10);
            final int thirdByte = Integer.parseInt(matcher.group(3), 10);
            final int forthByte = Integer.parseInt(matcher.group(4), 10);

            if (firstByte > 255 || secondByte > 255 || thirdByte > 255 || forthByte > 255) {
                return NotDeserialisableResult.create(7);
            }

            try {
                return DeserialisationResult.create(
                        Inet4Address.getByAddress(new byte[] { (byte) firstByte, (byte) secondByte, (byte) thirdByte, (byte) forthByte}),
                        address.length(),
                        buffer.hasRemaining());
            } catch (UnknownHostException e) {
                return NotDeserialisableResult.create(address.length());
            }
        }
        else {
            return NotDeserialisableResult.create(2);
        }
    }
}

File: proxy-protocol/src/main/clojure/com/mattunderscore/proxy/protocol/v1/Plaintextipv4deserialiser.clj

(ns com.mattunderscore.proxy.protocol.v1.Plaintextipv4deserialiser
  (import [com.mattunderscore.tcproxy.io.serialisation AbstractByteBufferDeserialiser NotDeserialisableResult NeedsMoreDataResult DeserialisationResult]
          [java.net.Inet4Address])
  (:gen-class
    :extends com.mattunderscore.tcproxy.io.serialisation.AbstractByteBufferDeserialiser))

(defn- is-group-separator [byte]
  (= byte (int \.)))

(defn- is-valid-byte [byte]
  (or (<= (int \0) byte (int \9))
      (is-group-separator byte)))

(defn- create-address-result-from-context [context remaining]
  (let [bytes (into-array Byte/TYPE (map (fn [group] (Integer/parseInt (clojure.string/join (map char group)))) (:groups context)))]
    (DeserialisationResult/create (java.net.Inet4Address/getByAddress bytes) (:processed context) (nil? (first remaining)))))

(defn- has-potential-address [context]
  (and
    (= 3 (count (:groups context)))
    (> (count (:pending context)) 0)))

(defn- is-grouping-invalid [byte context]
  (if (is-group-separator byte)
    (or
      ; Empty group
      (= 0 (count (:pending context)))
      ; Too many groups
      (= 4 (count (:groups context))))
    ; Too many characters in group
    (= 3 (count (:pending context)))))

(defn- append-pending-to-groups [context]
  (-> context
    (update-in [:groups] conj (:pending context))
    (assoc :pending [])))

(defn- process-next-byte [seq context]
  (if-let [byte (first seq)]
    (if (is-valid-byte byte)
      (if (is-grouping-invalid byte context)
        ; Group separator without pending group or too many groups
        (NotDeserialisableResult/create (+ (:processed context) 1))
        ; Accumulate the byte
        (if (is-group-separator byte)
          ; Pack new group
          (process-next-byte (rest seq) (-> context
                                          (update-in [:processed] + 1)
                                          (append-pending-to-groups)))
          ; Add to pending group
          (process-next-byte (rest seq) (-> context
                                            (update-in [:processed] + 1)
                                            (update-in [:pending] conj byte)))))

      (if (has-potential-address context)
        ; Create the address
        (create-address-result-from-context (append-pending-to-groups context) (rest seq))
        ; Ended early
        (NotDeserialisableResult/create (+ (:processed context) 1))))
    ; No more data
    (if (has-potential-address context)
      ; Create the address
      (create-address-result-from-context (-> context
                                              (update-in [:groups] conj (:pending context))
                                              (assoc :pending [])) (rest seq))
      ; Ended early
      (NeedsMoreDataResult/INSTANCE))))

(defn- read-ip-from-sequence [seq]
  (process-next-byte seq {:processed 0 :groups [] :pending []}))

(defn -doRead
  [this buffer]

  (let [byte-seq (repeatedly #(if (.hasRemaining buffer) (.get buffer) nil))]
    (read-ip-from-sequence byte-seq)))
